### PR TITLE
Task 7

### PR DESCRIPTION
**Pull request by femiarkh**
### 3. Self-evaluation: 6/6
- 1 - authorization-service is added to the repo, has correct basicAuthorizer lambda and correct serverless.yaml file
- 3 - import-service serverless.yaml file has authorizer configuration for the importProductsFile lambda. Request to the importProductsFile lambda should work only with correct authorization_token being decoded and checked by basicAuthorizer lambda. Response should be in 403 HTTP status if access is denied for this user (invalid authorization_token) and in 401 HTTP status if Authorization header is not provided.
- 5 - update client application to send Authorization: Basic authorization_token header on import. Client should get authorization_token value from browser localStorage https://developer.mozilla.org/ru/docs/Web/API/Window/localStorage authorization_token = localStorage.getItem('authorization_token')
- +1 - Client application should display alerts for the responses in 401 and 403 HTTP statuses. This behavior should be added to the nodejs-aws-fe-main/src/index.tsx file

### 4. Links:
- [Frontend PR](https://github.com/femiarkh/shop-react-redux-cloudfront/pull/5)
- [Deployed shop app](https://d2uxsq2yxa1sxw.cloudfront.net/)
Your test csv file may be something like:
```
title,description,price,count,image
bikeone,description,50,100,
biketwo,description,150,100,
```

### 5. How to test:
1. Open deployed front app, go to Manage products page and try to upload some `csv` file with new products (see example above).
2. Note `Unauthorized` alert and 401 error in console.
3. Open devtools --> Application tab --> Local storage --> https://d2uxsq2yxa1sxw.cloudfront.net/. Add a record there:
Key = `authorization_token`
Value = some random string
image
4. Reload the page and try to upload some `csv` file ones again.
5. Note `User is not authorized to access this resource with an explicit deny` alert and 403 error in console. 
6. Change value in local storage to `ZmVtaWFya2g6VEVTVF9QQVNTV09SRA==` (this is `femiarkh:TEST_PASSWORD` encoded in Base64). 
image
7. Reload the page and try to upload some `csv` file ones again.
8. There should be no error, and new products should appear in the catalog after some time.

### CODE REVIEW

File: authorization-service/src/functions/basicAuthorizer/handler.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import 'source-map-support/register';
+import { middyfy } from '@libs/lambda';
+import { Handler } from 'aws-lambda';
+
+const basicAuthorizer: Handler = async (event, _ctx, cb) => {
```
Comment (ragexe) [inline authorization-service/src/functions/basicAuthorizer/handler.ts:5]: Please note that in TS you can use generics for a better experience with the types
```
import {
  APIGatewayAuthorizerResult,
  APIGatewayTokenAuthorizerEvent,
  Handler,
} from 'aws-lambda';
const handler: Handler<
  APIGatewayTokenAuthorizerEvent,
  APIGatewayAuthorizerResult
> =
```

File: import-service/serverless.ts [hunk 2]
```diff
@@ -13,8 +15,11 @@ const serverlessConfiguration: AWS = {
       webpackConfig: './webpack.config.js',
       includeModules: true,
     },
+    authorizerArn: {
+      'Fn::ImportValue': 'basicAuthorizerQualifiedArn',
+    },
   },
-  plugins: ['serverless-webpack'],
+  plugins: ['serverless-webpack', 'serverless-dotenv-plugin'],
```
Comment (ragexe) [inline import-service/serverless.ts:22]: What was the reason to add 'dotenv-plugin' for import service?

**Review by ragexe — COMMENTED** [review]
## Evaluation criteria (each mark includes previous mark criteria)


- [x] * **1** - **authorization-service** is added to the repo, has correct **basicAuthorizer** lambda and correct **serverless.yaml** file
- [x] * **3** - **import-service** serverless.yaml file has authorizer configuration for the **importProductsFile** lambda. Request to the **importProductsFile** lambda should work only with correct **authorization_token** being decoded and checked by **basicAuthorizer** lambda. Response should be in 403 HTTP status if access is denied for this user (invalid **authorization_token**) and in 401 HTTP status if Authorization header is not provided.
- [x] * **5** - update client application to send Authorization: Basic **authorization_token** header on import. Client should get **authorization_token** value from browser localStorage https://developer.mozilla.org/ru/docs/Web/API/Window/localStorage
**authorization_token** = localStorage.getItem('**authorization_token**')
 
## Additional (optional) tasks - recommended for personal growth and further interviews, but this part would not be evaluated on cross-checking

- [ ] * **+1** - Client application should display alerts for the responses in 401 and 403 HTTP statuses. This behavior should be added to the **nodejs-aws-fe-main/src/index.tsx** file

image
image
> Response should be in 403 HTTP status if access is denied for this user (invalid **authorization_token**) and...

Actual result - **5 pts**

**Comment by femiarkh** [comment]
> ## Evaluation criteria (each mark includes previous mark criteria)
> * [x]  * **1** - **authorization-service** is added to the repo, has correct **basicAuthorizer** lambda and correct **serverless.yaml** file
> * [x]  * **3** - **import-service** serverless.yaml file has authorizer configuration for the **importProductsFile** lambda. Request to the **importProductsFile** lambda should work only with correct **authorization_token** being decoded and checked by **basicAuthorizer** lambda. Response should be in 403 HTTP status if access is denied for this user (invalid **authorization_token**) and in 401 HTTP status if Authorization header is not provided.
> * [x]  * **5** - update client application to send Authorization: Basic **authorization_token** header on import. Client should get **authorization_token** value from browser localStorage https://developer.mozilla.org/ru/docs/Web/API/Window/localStorage
>   **authorization_token** = localStorage.getItem('**authorization_token**')
> 
> ## Additional (optional) tasks - recommended for personal growth and further interviews, but this part would not be evaluated on cross-checking
> * [ ]  * **+1** - Client application should display alerts for the responses in 401 and 403 HTTP statuses. This behavior should be added to the **nodejs-aws-fe-main/src/index.tsx** file
> 
> image image
> 
> > Response should be in 403 HTTP status if access is denied for this user (invalid **authorization_token**) and...
> 
> Actual result - **5 pts**

You've probably forgot to reload the page (step 4). I've just checked with exact values you provided, all going smoothly. 
image

**Pull request merged**